### ISSUE TITLE
Tensorflow-gpu performance drop

### ISSUE THREAD

**Issue by TwoRice** [tensorflow/tensorflow#18297]
### System information
- **Have I written custom code**: No
- **OS Platform and Distribution**: Windows 10 64-bit 
- **TensorFlow installed from**: binary
- **TensorFlow version**: 1.6.0
- **Python version**: 3.6.3
- **CUDA/cuDNN version**: 9.0
- **GPU model and memory**: GeForce GTX 780, 3Gb

Using Keras in Python with tensorflow-gpu backend. Worked fine for weeks until a few days ago, when I have suffered a huge performance drop.

When Tensorflow is initialising, it all appears to work correctly and finds my GPU as normal. Output:

    2018-04-05 02:08:32.791893: I C:\tf_jenkins\workspace\rel-win\M\windows-gpu\PY\36\tensorflow\core\common_runtime\gpu\gpu_device.cc:1212] Found device 0 with properties: 
    name: GeForce GTX 780 major: 3 minor: 5 memoryClockRate(GHz): 1.0195
    pciBusID: 0000:01:00.0
    totalMemory: 3.00GiB freeMemory: 2.46GiB
    2018-04-05 02:08:32.792360: I C:\tf_jenkins\workspace\rel-win\M\windows-gpu\PY\36\tensorflow\core\common_runtime\gpu\gpu_device.cc:1312] Adding visible gpu devices: 0
    2018-04-05 02:08:33.132555: I C:\tf_jenkins\workspace\rel-win\M\windows-gpu\PY\36\tensorflow\core\common_runtime\gpu\gpu_device.cc:993] Creating TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 2187 MB memory) -> physical GPU (device: 0, name: GeForce GTX 780, pci bus id: 0000:01:00.0, compute capability: 3.5)

But whereas before with exactly the same model on the same data, each epoch took about 2-3 seconds, they now take ~17 seconds.

I had a look in Task Manager, and my IDE shows to be using GPU Engine - "GPU 0 - Copy". Also at the beginning of each epoch, the GPU will go under ~70% load for about a second, but then the load switches over to my CPU and Memory for the remaining 15 seconds or so and the GPU goes back down to its idle load around 2%.

**Comment by calledbymountains** [comment]
This is consistent with my experiments. As of commit `5d33c1e49178aedbb459da7ce58eca710102c06b` , the performance drop over 25600 images of imagenet when training densenet265 is about 17 seconds.

**Comment by yaroslavvb** [comment]
The official benchmarks have not been affected as you can see [here](https://benchmarks-dot-tensorflow-testing.appspot.com/)

So the question is what you are doing different from official pipelines.

One possibility is if you are using feed_dict, which is not well tested by those benchmarks. Upgrading past TensorFlow 1.6 changes alignment requirement which will lead to slowdown if you use feed_dict, see https://github.com/tensorflow/tensorflow/issues/17233

**Comment by TwoRice** [comment]
Not 100% sure what is being used because I'm using Keras. 
Would I receive a warning if feed_dict was being used?
I do get this warning, however

`WARNING:tensorflow:From C:\Python36\lib\site-packages\keras\backend\tensorflow_backend.py:1259: calling reduce_prod (from tensorflow.python.ops.math_ops) with keep_dims is deprecated and will be removed in a future version.
Instructions for updating:
keep_dims is deprecated, use keepdims instead
`

If that might have anything to do with it.

**Comment by yaroslavvb** [comment]
I guess the best way to debug would be to look at timeline of a single step (fast) of old version, and compare it to timeline of single step (slow) in new version, and see what changed

**Comment by TwoRice** [comment]
Like I said, I'm using the same data and model as before, and I haven't updated any of my libraries.
Although I guess _something_ must have changed that's caused this performance drop, but at the moment, I have no idea how to get my fast version back.

**Comment by yaroslavvb** [comment]
Are you sure this is not a user error? Without a reproducible case or some kind of narrowing down of the regression, this report is not very useful since it's hard to make progress on it

**Comment by ppwwyyxx** [comment]
@TwoRice if you're using Keras and you don't know whether you're using feed_dict or not, you're probably using feed_dict in the critical path (because it's a Keras default). 
feed_dict is slow probably since the first day of tensorflow. It gets slower in 1.6 as @yaroslavvb mentioned. Maybe related to your issue.

**Comment by reedwm** [comment]
/CC @tatianashp, do you think this could be related to #17233?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by jlopezpena** [comment]
Don't know if it is due to the same reason, but I am also seeing some regressions in the 1.4 - 1.5 and 1.5 - 1.6 updates, as reported in #19027 (with reproducible snippet).
In my example, the slowdown happens in CPU computations, and the code becomes 100x slower. Not sure where it is due to tensorflow, to keras, or to anaconda compilation flags.

**Comment by yaroslavvb** [comment]
@jlopezpena try running the following before running your script

```
sudo apt-get install google-perftools
export LD_PRELOAD="/usr/lib/libtcmalloc.so.4" 
```

This changes alignment of default numpy array and fixes the slowdown due to feed_dict introduced in 1.6

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 60 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 75 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 90 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 105 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 120 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 135 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 150 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 165 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 180 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by mohantym** [comment]
Hi @TwoRice !
It seems you are using older versions(1.x versions) of Tensorflow which is not supported any more. Have you tried [latest version](https://www.tensorflow.org/install/gpu)(2.7) yet? Thanks!
